pkg/buildkit/builder.go

```diff
@@ -1,42 +1,171 @@
+		shellExecutor   executor.ShellExecutor
+		contextProvider ContextProvider
+	)
+
+	if localOnly {
+		shellExecutor = exec.NewShellExecutor(workingDir, nil)
+		contextProvider = NewLocalContextProvider()
+	} else {
+		k8sExecutor, err = createBuildkitKubernetesExecutor()
+		if err != nil {
+			logger.Fatalf("cannot create buidkit kubernetes executor: %v", err)
+		}
+
+		awsCfg, err := config.LoadDefaultConfig(context.Background(),
+			config.WithRegion(cfg.Context.S3.Region))
+		if err != nil {
+			logger.Fatalf("cannot load AWS config: %v", err)
+		}
+		s3 := NewS3Uploader(awsCfg, cfg.Context.S3.Bucket)
+		contextProvider = NewRemoteContextProvider(s3)
+	}
+
 	return &Builder{
-		executor:       exec,
-		buildctlBinary: binary,
+		bkShellExecutor: bkShellExecutor{
+			shellExecutor,
+			binary,
+		},
+		bkKubernetesExecutor: bkKubernetesExecutor{
+			KubernetesExecutor: k8sExecutor,
+			buildctlBinary:     binary,
+			dockerConfigSecret: cfg.Executor.Kubernetes.DockerConfigSecret,
+			podConfig: k8sutils.PodConfig{
+				Namespace:     cfg.Executor.Kubernetes.Namespace,
+				NameGenerator: k8sutils.UniquePodName("buildkit-dib"),
+				Labels: map[string]string{
+					"app.kubernetes.io/managed-by": "dib",
+				},
+				Image:            cfg.Executor.Kubernetes.Image,
+				ImagePullSecrets: cfg.Executor.Kubernetes.ImagePullSecrets,
+				EnvSecrets:       cfg.Executor.Kubernetes.EnvSecrets,
+				Env: map[string]string{
+					"AWS_REGION": cfg.Context.S3.Region,
+					"container":  "kube", // Fix for https://github.com/GoogleContainerTools/kaniko/issues/1542
+				},
+				PodOverride:       cfg.Executor.Kubernetes.PodTemplateOverride,
+				ContainerOverride: cfg.Executor.Kubernetes.ContainerOverride,
+			},
+		},
+		contextProvider: contextProvider,
 	}, nil
 }
 
 // Build the image using the Buildkit backend.
 func (b Builder) Build(opts types.ImageBuilderOpts) error {
+	var err error
+
+	opts.Context, err = b.contextProvider.PrepareContext(opts)
+	if err != nil {
+		return fmt.Errorf("cannot prepare buildkit build context: %w", err)
+	}
 	buildctlArgs, err := generateBuildctlArgs(opts)
 	if err != nil {
 		return err
 	}
-
-	// Currently, we only support writing to stdout. In the future, this can be improved to support custom stdio.
-	if err := b.executor.ExecuteStdout(b.buildctlBinary, buildctlArgs...); err != nil {
-		return err
+	// `shellExecutor` or `kubernetesExecutor` are mutually exclusive.
+	if b.bkShellExecutor.shellExecutor != nil {
+		//nolint:lll
+		if err := b.bkShellExecutor.shellExecutor.ExecuteStdout(b.bkShellExecutor.buildctlBinary, buildctlArgs...); err != nil {
+			return err
+		}
+	} else {
+		pod, err := buildPod(b.bkKubernetesExecutor.dockerConfigSecret, b.bkKubernetesExecutor.podConfig, buildctlArgs)
+		if err != nil {
+			return err
+		}
+		//nolint:lll
+		if err := b.bkKubernetesExecutor.KubernetesExecutor.ApplyWithWriters(context.Background(), opts.LogOutput, opts.LogOutput, pod, "buildkit"); err != nil {
```

Comment:
To make it more readable maybe?

```suggestion
		if err := b.k8sExecutor.K8sExecutor.ApplyWithWriters(context.Background(), opts.LogOutput, opts.LogOutput, pod, "buildkit"); err != nil {
```

Reply:
hmm I find that `Kubernetes` is more expressive than `k8s`.

Reply:
Alright 👌